pkg/operator/watcher/filewatcher.go

```diff
@@ -0,0 +1,146 @@
+package watcher
+
+import (
+	"fmt"
+	"io/ioutil"
+	"sync"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+	"gopkg.in/fsnotify.v1"
+)
+
+var log = logf.Logger.WithName("filewatcher")
+
+// FileWatcher watches a file for changes.
+type FileWatcher struct {
+	sync.Mutex
+	file        string
+	currentData []byte
+	watcher     *fsnotify.Watcher
+}
+
+// New returns a new FileWatcher watching the given file.
+func New(file string) (*FileWatcher, error) {
+	var err error
+
+	cw := &FileWatcher{
+		file: file,
+	}
+
+	// Initial read of file.
+	if err := cw.ReadFile(); err != nil {
+		return nil, err
+	}
+
+	cw.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	return cw, nil
+}
+
+// GetFile fetches the file being watched by FileWatcher.
+func (fw *FileWatcher) GetFile() string {
+	fw.Lock()
+	defer fw.Unlock()
+	return fw.file
+}
+
+// GetFileData fetches the currently loaded file data of FileWatcher.
+func (fw *FileWatcher) GetFileData() []byte {
+	fw.Lock()
+	defer fw.Unlock()
+	return fw.currentData
+}
+
+// Start starts the FileWatcher.
+func (fw *FileWatcher) Start(stopCh <-chan struct{}) error {
+	if err := fw.watcher.Add(fw.file); err != nil {
+		return err
+	}
+
+	go fw.Watch()
+
+	log.Info("Starting file watcher")
+
+	// Block until the stop channel is closed.
+	<-stopCh
+
+	return fw.watcher.Close()
+}
+
+// Watch reads events from the watcher's channel and reacts to changes.
+func (fw *FileWatcher) Watch() {
+	for {
+		select {
+		case event, ok := <-fw.watcher.Events:
+			// Channel is closed.
+			if !ok {
+				return
+			}
+
+			fw.handleEvent(event)
+
+		case err, ok := <-fw.watcher.Errors:
+			// Channel is closed.
+			if !ok {
+				return
+			}
+
+			log.Error(err, "file watch error")
+		}
+	}
+}
+
+// ReadFile reads the watched file from disk, parses the file,
+// and updates FileWatcher current data.
+func (fw *FileWatcher) ReadFile() error {
+	data, err := ioutil.ReadFile(fw.file)
+	switch {
+	case err != nil:
+		return fmt.Errorf("failed to read file %s: %v", fw.file, err)
+	case len(data) == 0:
+		return fmt.Errorf("file %s contains no currentData", fw.file)
+	}
+
+	fw.Lock()
+	fw.currentData = data
+	fw.Unlock()
+
+	log.Info("updated file watcher current data")
+
+	return nil
+}
+
+// handleEvent filters events, re-adds and re-reads the watched file
+// if removed.
+func (fw *FileWatcher) handleEvent(event fsnotify.Event) {
+	if !(isWrite(event) || isRemove(event) || isCreate(event)) {
+		return
+	}
+
+	log.Info("watched file change", "event", event)
+
+	if isRemove(event) {
+		if err := fw.watcher.Add(event.Name); err != nil {
+			log.Error(err, "error re-watching file %s", fw.file)
+		}
```

Comment:
Return here?  Or do we want to try to read the file after receiving a remove event?

Reply:
This is what the watcher sees when the trusted ca bundle configmap is updated (i.e. a new ca cert is added):
`2019-12-09T19:55:31.185Z	INFO	operator.filewatcher	watcher/filewatcher.go:113	watched file change	{"event": "\"/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem\": REMOVE"}`